lasagne/layers/merge.py

```diff
@@ -312,9 +327,20 @@ def __init__(self, incomings, merge_function, cropping=None, **kwargs):
 
     def get_output_shape_for(self, input_shapes):
         input_shapes = autocrop_array_shapes(input_shapes, self.cropping)
-        if any(shape != input_shapes[0] for shape in input_shapes):
+        # Infer as much of the shape as possible, avoiding Nones
+        output_shape = tuple(next((s[i] for s in input_shapes
+                                   if s[i] is not None), None)
+                             for i in range(len(input_shapes[0])))
+
```

Comment:
Good idea about swapping things with `zip` (although `shape` should be `sizes` then, as it's all sizes for a particular axis)! If I saw a `for` loop I'd be wondering why it's not a list comprehension. It would get quite long:
```python
output_shape = []
ndim = len(input_shapes[0])
for i in range(ndim):
    size = None
    for shape in input_shapes:
        if shape[i] is not None:
            size = shape[i]
            break
    output_shape.append(size)
```
I find it more difficult to read this, and I'd feel bad about having this twice. I'll change the code according to your suggestion and add a second comment line stating that we're simply grabbing the first non-None-size per dimension.